### PR TITLE
Use queries for the HIR map

### PR DESCRIPTION
**Pull request by Zoxc**
r? @eddyb cc @michaelwoerister

### CODE REVIEW

**Comment by Zoxc** [comment]
@rust-timer build 534242c9670aa413fc779b7d3c4f70865c38c2ae

**Comment by Zoxc** [comment]
@rust-timer build c6c89b70fa50afc9968591c5b409fd70b2ca1f2c

**Comment by Zoxc** [comment]
This PR makes the `hir::Map` type into a wrapper type which just calls these new queries:
```rust
query index_hir(_: CrateNum) -> &'tcx map::IndexedHir<'tcx> {
    eval_always
    no_hash
    desc { "index HIR" }
}

query hir_module_items(key: DefId) -> &'tcx hir::ModuleItems {
    eval_always
}

query hir_owner(key: DefId) -> &'tcx HirOwner<'tcx> {
    eval_always
}

query hir_owner_items(key: DefId) -> &'tcx HirOwnerItems<'tcx> {
    eval_always
}

query all_local_trait_impls(key: CrateNum) -> &'tcx BTreeMap<DefId, Vec<hir::HirId>> {
    desc { "local trait impls" }
}
```

The previous fields of `hir::Map` is now stored in `IndexedHir` which the `index_hir` query will calculate. The `hir_module_items`, `hir_owner`, `hir_owner_items` and `all_local_trait_impls` queries will just return already computed information and are just used to narrow dependencies.

New types:
```rust
pub struct IndexedHir<'hir> {
    /// The SVH of the local crate.
    pub crate_hash: Svh,

    pub(super) owner_map: FxHashMap<DefIndex, &'hir HirOwner<'hir>>,
    pub(super) owner_items_map: FxHashMap<DefIndex, &'hir HirOwnerItems<'hir>>,

    /// The reverse mapping of `node_to_hir_id`.
    pub(super) hir_to_node_id: FxHashMap<HirId, NodeId>,
}

pub struct HirOwner<'tcx> {
    parent: HirId,
    node: Node<'tcx>,
}

pub struct HirItem<'tcx> {
    parent: ItemLocalId,
    node: Node<'tcx>,
}

pub struct HirOwnerItems<'tcx> {
    owner: Node<'tcx>,
    items: IndexVec<ItemLocalId, Option<HirItem<'tcx>>>,
    bodies: FxHashMap<ItemLocalId, &'tcx Body<'tcx>>,
}
```

cc @rust-lang/compiler

File: src/librustc/hir/map/collector.rs
Comment (eddyb) [inline src/librustc/hir/map/collector.rs]: _**n i c e**_

File: src/librustc_hir/hir.rs
Comment (eddyb) [inline src/librustc_hir/hir.rs]: Would be nice if we could get away with an outright module `Item`.

**Review by eddyb — APPROVED** [review]
LGTM, and ftr this matches some of my own plans (and makes them easier).

That said, it would be nice if someone else could also look over it.
(Also, I'm not sure where we're at with perf, and this PR might also need some design discussion?)

**Comment by michaelwoerister** [comment]
Thanks for the PR, @Zoxc. Here are some thoughts:

- I _think_ this looks very promising and is along the lines of what was discussed in the end-to-end queries design meeting. I can't quite tell yet. It would be great to have a bit more documentation on what the new types introduced here are exactly, e.g.:
  - What is `IndexedHir`? I'd guess the entire HIR after lowering + lookup tables?
  - What is a `HirOwner`? An "item-like"? The thing corresponding to a `DepNode::Hir`?
  - What is a `HirItem`? A child of an owner, like an expression, params, etc?
  - What `HirOwnerItems`? A local lookup table for items of a `HirOwner`?
- Performance looks quite bad at the moment. We'll need to find a way fix this before going forward. Some superficial browsing of the detailed perf.rlo tables (e.g. https://perf.rust-lang.org/detailed-query.html?commit=c6c89b70fa50afc9968591c5b409fd70b2ca1f2c&base_commit=64ea639c12df0594dd891b1ba0b439c8c5eacd83&benchmark=piston-image-check&run_name=patched%20incremental:%20println) seems to suggest that there is more invalidation going on (that the number of calls to things like `typeck_tables_of` increases). Not sure if that is due to bug fixes or just a different configuration of red-green boundaries. Do we decrease dep-node granularity for HIR-related things with this setup?
- Does this get rid of all custom dep-tracking we do for the HIR map? (That would be awesome!)
- Does this hash the HIR twice in order to compute the SVH? (I have some ideas around a low-tech solution for memoizing Fingerprints that is mainly aimed at avoiding re-hashing upstream MIR, but which also could help with the SVH problem).

**Comment by eddyb** [comment]
> What is a HirOwner? An "item-like"? The thing corresponding to a DepNode::Hir?

Assuming @Zoxc is using the same terminology as I am, "HIR owner" is replacing the outdated and/or inaccurate "item-like", and eventually might be replaced itself by "HIR def" (but that's farther off).

**Comment by Zoxc** [comment]
@rust-timer build 090412221c6300b5828f61d6f39e2777b94214ab

**Comment by Zoxc** [comment]
> * What is `IndexedHir`? I'd guess the entire HIR after lowering + lookup tables?

It has the crate hash, the lookup tables for `hir_owner` and `hir_owner_items` and also `hir_to_node_id`. It basically store the stuff that was in `Map`.

> * What is a `HirOwner`? An "item-like"? The thing corresponding to a `DepNode::Hir`?

It is an item-like or a macro def. It corresponds to `DepNode::Hir`.

> * What is a `HirItem`? A child of an owner, like an expression, params, etc?

Yeah.

> * What `HirOwnerItems`? A local lookup table for items of a `HirOwner`?

It's a lookup table for items (without a `DefIndex`) and bodies. It corresponds to `DepNode::HirBody`.


> Not sure if that is due to bug fixes or just a different configuration of red-green boundaries. Do we decrease dep-node granularity for HIR-related things with this setup?

The dep-node granularity is the same, expect for `hir_module_items` which is slightly finer.

I looked over the existing HIR map and [fixed the bugs I found](https://github.com/rust-lang/rust/pull/69015). That seems to result in [similar regressions](https://perf.rust-lang.org/compare.html?start=840bdc349d2885a5173269b653025192969cfebc&end=f84d9860696824a4789e785ae1a1a4e96ee55257).

> * Does this get rid of all custom dep-tracking we do for the HIR map? (That would be awesome!)

It does.

> * Does this hash the HIR twice in order to compute the SVH?

It hashed the HIR thrice. I put `while_hashing_hir_bodies(false, ...` on the wrong type. I fixed that and also made `HirOwnerItems` store a `Fingerprint` calculated by the collector. So now this hashes HIR as much as `master`.

**Comment by Zoxc** [comment]
This PR seems to trigger https://github.com/rust-lang/rust/issues/62649 / `issue-62649-path-collisions-happen.rs` again, which means we probably didn't fix the root cause.

**Comment by Zoxc** [comment]
The problem seems to be that `as_local_hir_id` called from `type_of` is untracked and reveals the `HirId` without adding a dependency. When `Something::foo` is changed from a method to a field, this `HirId` will change without causing `type_of` to be re-executed.

**Comment by Zoxc** [comment]
@rust-timer build 74fce5097f923d83640aaba1861c89fd3ad3d898

**Comment by Zoxc** [comment]
It seems like https://github.com/rust-lang/rust/pull/68693 helped a bit with the performance here.

**Comment by Zoxc** [comment]
Now perf is looking more reasonable.

File: src/librustc/dep_graph/graph.rs
Comment (michaelwoerister) [inline src/librustc/dep_graph/graph.rs]: I see how this is still relevant for `DepKind::CrateMetadata` but I'm surprised that it's needed for `DepKind::hir_owner` and `DepKind::hir_owner_items`.

**Comment by michaelwoerister** [comment]
> Now perf is looking more reasonable.

Indeed. And it looks like no unsafe caching outside of dependency tracking was needed. That is great news!

File: src/librustc/dep_graph/graph.rs
Comment (Zoxc) [reply src/librustc/dep_graph/graph.rs]: Given that `hir_owner` is just a rename of `Hir` and `hir_owner_items` is just a rename of `HirBody`, you can follow the [explanation here](https://github.com/rust-lang/rust/issues/62649#issuecomment-589018098) to see how this has the same problem as `Hir` and `HirBody`. Also `Hir(Something)` will be replaced by `hir_owner_items(Something)` there, but that doesn't affect anything.

**Comment by petrochenkov** [comment]
>ping in https://github.com/rust-lang/rust/pull/68944#issuecomment-583866038

[triagebot] I don't understand what exactly this PR does and why, there seems to be no documentation to figure that out, and I don't have time to investigate this by myself.

File: src/librustc/dep_graph/graph.rs
Comment (michaelwoerister) [reply src/librustc/dep_graph/graph.rs]: I had a chance to test this and I think the explanation in https://github.com/rust-lang/rust/issues/62649#issuecomment-589082343 basically applies here:

- `Something::foo` is a valid query key for `hir_owner` in `rpass1` but in `rpass2` it's not a valid key anymore. 
- The red-green algorithm does not detect this because `RecoverKey::recover()` returns `Some(_)` instead of `None`. The algorithm thus goes ahead with forcing `DepNode::hir_owner(Something::foo)` and ICEs.

A short-term fix for this would be to have a newtype `struct HirOwnerDefId(DefId)` that is a `DefId` with the additional invariant that it must correspond to a hir-owner. Then `<HirOwnerDefId as RecoverKey>::recover()` could apply the same check as done [here](https://github.com/rust-lang/rust/blob/2cb0b8582ebbf9784db9cec06fff517badbf4553/src/librustc/dep_graph/graph.rs#L918-L921) and return `None` if it doesn't hold. Most uses of `HirOwnerDefId` should stay hidden behind the hir map wrapper, so this shouldn't be too disruptive.

Another workaround would be to diversify the `DefPathData` variants again so that a field and a method would not have the same `DefPath` anymore. Either fixes would be acceptable solutions for unblocking this PR, in my opinion. 

I think one way to fix the problem at the conceptual level would be to add the invariant that `RecoverKey::recover()` must only return `Some(_)` if the recovered key is actually a valid query-key for the given query. That is, the result of `RecoverKey::recover()` would need to depend on the query, not just on the type of the query-key. It would then be able to do query-specific runtime checks about key validity.

Comment (Zoxc) [reply src/librustc/dep_graph/graph.rs]: This doesn't block this PR though. It's just https://github.com/rust-lang/rust/pull/68289 ported to it. There's no further problems caused by this PR compared to master.

It seems like you're trying to fix https://github.com/rust-lang/rust/issues/62649 here. I mentioned the `DefPathData` solution in a [comment](https://github.com/rust-lang/rust/issues/62649#issuecomment-585355880)) there. I want to check the performance of making `as_local_hir_id` a query, but that is blocked on this PR.

Comment (michaelwoerister) [reply src/librustc/dep_graph/graph.rs]: Well, I'm fine with merging this if you mention #62649 in the FIXME comment above.

**Comment by michaelwoerister** [comment]
I'm in favor of merging this. It removes most special casing around HIR dependency tracking, making the HIR map query-based. 

It would be great to add doc comments describing the new queries and that they are not supposed to be used directly. It also would be nice to squash the some of the earlier commits that contain lots of things that don't make it to the final diff. 

Heads up, @rust-lang/compiler-contributors, this [slightly regresses performance](https://perf.rust-lang.org/compare.html?start=6af388b25050bca26710be7e4030e17bf6d8d2f7&end=264692ca7a852787e74bb25c42ff476b2fc14d96&stat=instructions%3Au) for some of the check builds. However, the regression is small and mostly in synthetic test cases. Debug- and opt-builds are not noticeably affected. 

I'll leave it to @eddyb to give the final r+.

File: src/librustc_hir/intravisit.rs
Comment (Centril) [inline src/librustc_hir/intravisit.rs]: I would prefer if we kept the reference here so that `M: ?Sized` was valid, specifically such that `dyn Map<'v>` was valid. (See https://github.com/rust-lang/rust/pull/69920.)

Comment (Zoxc) [reply src/librustc_hir/intravisit.rs]: I think you should be able to use `type Map = &'a dyn Map<'v>`.

Comment (Centril) [reply src/librustc_hir/intravisit.rs]: That would work I think.

File: src/librustc/dep_graph/graph.rs
Comment (Zoxc) [reply src/librustc/dep_graph/graph.rs]: I've added a reference to that issue.

**Comment by Zoxc** [comment]
Note that the [last perf run](https://perf.rust-lang.org/compare.html?start=6af388b25050bca26710be7e4030e17bf6d8d2f7&end=264692ca7a852787e74bb25c42ff476b2fc14d96) included the use of `LocalDenseDefIdCache` introduced in https://github.com/rust-lang/rust/pull/69303. That's no longer part of this PR as I want to measure the performance impact of that separately in a follow up PR.


@bors rollup=never

File: src/librustc/query/mod.rs
Comment (eddyb) [inline src/librustc/query/mod.rs]: Long-term, we should probably turn this (and other similar things if there are any) into a regular query by iterating through all HIR items or something.

**Review by eddyb — APPROVED** [review]

File: src/librustc/dep_graph/dep_node.rs
Comment (Aaron1011) [inline src/librustc/dep_graph/dep_node.rs]: Is this supposed to be a  `?`?

**Comment by JohnTitor** [comment]
Let's wake bors.
@bors retry

File: src/librustc/dep_graph/dep_node.rs
Comment (Zoxc) [reply src/librustc/dep_graph/dep_node.rs]: No, but this PR removed all the suitable examples...

Comment (Zoxc) [reply src/librustc/dep_graph/dep_node.rs]: I've removed the `?` and instead just don't use an example here.

**Comment by Zoxc** [comment]
Perf result: https://perf.rust-lang.org/compare.html?start=e0f5df017368dc3f7cb458fc6d5a5e0420e1d2e5&end=45ebd5808afd3df7ba842797c0fcd4447ddf30fb&stat=instructions%3Au

**Comment by bjorn3** [comment]
`mir_built` became 30-40% slower.

**Pull request merged**
